adapters/rise/rise.go

```diff
@@ -0,0 +1,134 @@
+	_ *adapters.ExtraRequestInfo) (
+	requestsToBidder []*adapters.RequestData,
+	errs []error,
+) {
+	publisherID, err := a.extractPublisherID(openRTBRequest)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("extractPublisherID: %w", err))
+		return nil, errs
+	}
+
+	openRTBRequestJSON, err := json.Marshal(openRTBRequest)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("marshal bidRequest: %w", err))
+		return nil, errs
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+
+	return append(requestsToBidder, &adapters.RequestData{
+		Method:  http.MethodPost,
+		Uri:     a.endpointURL + "?publisher_id=" + publisherID,
+		Body:    openRTBRequestJSON,
+		Headers: headers,
+	}), nil
+}
+
+func (a *adapter) extractPublisherID(openRTBRequest *openrtb2.BidRequest) (string, error) {
+	var (
+		bidderExt   adapters.ExtImpBidder
+		impExt      openrtb_ext.ImpExtRise
+		publisherID string
+	)
+
+	for _, imp := range openRTBRequest.Imp {
+		if err := json.Unmarshal(imp.Ext, &bidderExt); err == nil {
+			if err = json.Unmarshal(bidderExt.Bidder, &impExt); err == nil && impExt.PublisherID != "" {
+				publisherID = strings.TrimSpace(impExt.PublisherID)
+				break
+			}
+		}
+	}
+	if publisherID == "" {
+		return "", errors.New("no publisherID supplied")
+	}
+
+	return publisherID, nil
+}
+
+// MakeBids unpacks the server's response into Bids.
+func (a *adapter) MakeBids(
+	request *openrtb2.BidRequest,
+	_ *adapters.RequestData,
+	responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if responseData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if responseData.StatusCode == http.StatusBadRequest {
+		err := &errortypes.BadInput{
+			Message: "Unexpected status code: 400. Bad request from publisher. Run with request.debug = 1 for more info.",
+		}
+		return nil, []error{err}
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+
+	for _, seatBid := range response.SeatBid {
+		for i, bid := range seatBid.Bid {
+			bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+				Bid:     &seatBid.Bid[i],
+				BidType: getMediaTypeForBid(bid.ImpID, request.Imp),
+			})
+		}
+	}
+
+	return bidResponse, nil
+}
+
+func getMediaTypeForBid(impID string, imps []openrtb2.Imp) openrtb_ext.BidType {
+	for _, imp := range imps {
+		if imp.ID == impID {
+			if imp.Banner != nil {
+				return openrtb_ext.BidTypeBanner
+			} else if imp.Video != nil {
+				return openrtb_ext.BidTypeVideo
+			}
```

Comment:
Warning - This is an anti-pattern. This code assumes that for multi-format bid request, mediatype will be set to type banner by default. I would recommend to set the mediaType based on bidder response. Your bidder should be able to set openrtb2.Bid.Mtype field and use that field here to get mediaTypeForImp. Even if your bidder does not support multi-format request, I would highly encourage you to make this change. [Here is an example.](https://github.com/prebid/prebid-server/blob/6b616196605030f568dae4a727d1a658de929417/adapters/ix/ix.go#L246-L247)